File: 2022/Day8.py
#! /usr/bin/env python
# pylint: disable=unused-wildcard-import
from util import *

def parse(inp):
    return grid(inp)

def p1(inp):
    w, h = inp.shape
    g = inp
    c = 0
    for i in range(w):
        for j in range(h):
            v = g[i,j]
            G = g - v
            xs = [G[i, j+1:], G[i, :j], G[i+1:, j], G[:i, j]]
            if any(all(x < 0) for x in xs): c+=1
    return c

def p2(inp):
    w, h = inp.shape
    g = inp
    m = 0
    for i in range(1, w-1):
        for j in range(1, h-1):
            v = g[i,j]
            xs = [g[i, j+1:], g[i, :j][::-1], g[i+1:, j], g[:i, j][::-1]]
            f = lambda x: len(list(it.takewhile(lambda a: a < v, x)))
            ys = [fx if fx == len(x) else 1+fx for x, fx in [(x, f(x)) for x in xs]]
            m = max(m, prod(ys))
    return m

def main():
    inp = parse(data())
    print(f"Solution for part 1:\n{p1(inp)}")
    print(f"Solution for part 2:\n{p2(inp)}")

if __name__ == "__main__":
    main()

# Solution part 1: 1814
# Solution part 2: 330786

File: 2022/Day8.hs
#! /usr/bin/env -S runhaskell
import Data.List
import Data.Maybe (fromMaybe)
import Control.Applicative

import Util

type Input = [[Int]]
type Output = Int

parse :: String -> Maybe Input
parse = grid

apply4Way :: ([a] -> [b]) -> (b -> b -> b) -> [[a]] -> [[b]]
apply4Way transformLine combineOp = ((map (uncurry combine) . ) . zip) <$> transformGrid <*> (transpose . transformGrid . transpose)
  where
    transformGrid = map (combine <$> transformLine <*> reverse . transformLine . reverse)
    combine = (map (uncurry combineOp) . ) . zip

part1 :: Input -> Output
part1 = sumBGrid . apply4Way transformLine (||)
  where transformLine = fst . foldr (\x (a, hi) -> ((hi < x):a, max hi x)) ([], -1)

part2 :: Input -> Output
part2 = (reducerGrid max) . apply4Way transformLine (*)
  where transformLine = map (\x -> fromMaybe ((length x) - 1) $ findIndex (>= head x) (head x - 1 : tail x)) . init . tails

main :: IO()
main = do
    f <- readFile("8.in")
    let input = parse f
    case input of
      Nothing -> putStrLn "Unable to read input."
      Just i -> do
        putStr "Solution for part 1: "
        putStrLn . show $ part1 i
        putStr "Solution for part 2: "
        putStrLn . show $ part2 i

-- Solution part 1: 1814
-- Solution part 2: 330786
